File: android/app/src/main/java/com/turntable/barcodescanner/SearchHistoryStore.kt
package com.turntable.barcodescanner

import android.content.Context
import org.json.JSONArray
import org.json.JSONObject

data class SearchHistoryEntry(
    val timestampMs: Long,
    val barcode: String,
    val title: String,
    val coverUrl: String?,
)

object SearchHistoryStore {
    private const val KEY_HISTORY_JSON = "search_history_json"
    private const val MAX_ENTRIES = 100

    fun add(context: Context, barcode: String, title: String, coverUrl: String?) {
        val cleanBarcode = barcode.trim()
        val cleanTitle = title.trim()
        val cleanCover = coverUrl?.trim()?.takeIf { it.isNotBlank() }
        if (cleanBarcode.isEmpty() && cleanTitle.isEmpty()) return
        val current = getAll(context).toMutableList()
        current.add(
            0,
            SearchHistoryEntry(
                timestampMs = System.currentTimeMillis(),
                barcode = cleanBarcode,
                title = cleanTitle,
                coverUrl = cleanCover,
            ),
        )
        save(context, current.take(MAX_ENTRIES))
    }

    fun getAll(context: Context): List<SearchHistoryEntry> {
        val raw = context
            .getSharedPreferences(SearchPrefs.PREFS_NAME, Context.MODE_PRIVATE)
            .getString(KEY_HISTORY_JSON, null)
            ?: return emptyList()
        return try {
            val arr = JSONArray(raw)
            buildList {
                for (i in 0 until arr.length()) {
                    val o = arr.optJSONObject(i) ?: continue
                    add(
                        SearchHistoryEntry(
                            timestampMs = o.optLong("timestampMs", 0L),
                            barcode = o.optString("barcode", ""),
                            // Backward-compatible migration from older `query` key.
                            title = o.optString("title", o.optString("query", "")),
                            coverUrl = o.optString("coverUrl", "").takeIf { it.isNotBlank() },
                        ),
                    )
                }
            }
        } catch (_: Exception) {
            emptyList()
        }
    }

    private fun save(context: Context, entries: List<SearchHistoryEntry>) {
        val arr = JSONArray()
        for (e in entries) {
            arr.put(
                JSONObject().apply {
                    put("timestampMs", e.timestampMs)
                    put("barcode", e.barcode)
                    put("title", e.title)
                    put("coverUrl", e.coverUrl ?: "")
                },
            )
        }
        context.getSharedPreferences(SearchPrefs.PREFS_NAME, Context.MODE_PRIVATE)
            .edit()
            .putString(KEY_HISTORY_JSON, arr.toString())
            .apply()
    }
}

File: ios/turnTable/SearchHistoryStore.swift
import Foundation

struct SearchHistoryEntry: Codable {
    let timestampMs: Int64
    let barcode: String
    let title: String
    let coverUrl: String?
}

/// Mirrors Android `SearchHistoryStore` — JSON in `search_prefs` under `search_history_json`.
enum SearchHistoryStore {
    private static let key = "search_history_json"
    private static let maxEntries = 100

    private static var prefs: UserDefaults {
        UserDefaults(suiteName: "search_prefs") ?? .standard
    }

    static func add(barcode: String, title: String, coverUrl: String?) {
        let cleanBarcode = barcode.trimmingCharacters(in: .whitespacesAndNewlines)
        let cleanTitle = title.trimmingCharacters(in: .whitespacesAndNewlines)
        let cleanCover = coverUrl?.trimmingCharacters(in: .whitespacesAndNewlines).nilIfEmpty
        if cleanBarcode.isEmpty && cleanTitle.isEmpty { return }
        var current = getAll()
        current.insert(
            SearchHistoryEntry(
                timestampMs: Int64(Date().timeIntervalSince1970 * 1000),
                barcode: cleanBarcode,
                title: cleanTitle,
                coverUrl: cleanCover
            ),
            at: 0
        )
        if current.count > maxEntries {
            current = Array(current.prefix(maxEntries))
        }
        save(current)
    }

    static func getAll() -> [SearchHistoryEntry] {
        guard let raw = prefs.string(forKey: key), let data = raw.data(using: .utf8) else { return [] }
        return (try? JSONDecoder().decode([SearchHistoryEntry].self, from: data)) ?? []
    }

    private static func save(_ entries: [SearchHistoryEntry]) {
        guard let data = try? JSONEncoder().encode(entries), let s = String(data: data, encoding: .utf8) else { return }
        prefs.set(s, forKey: key)
    }
}

private extension String {
    var nilIfEmpty: String? {
        let t = trimmingCharacters(in: .whitespacesAndNewlines)
        return t.isEmpty ? nil : self
    }
}
